### PR TITLE
METRON-364: Preserve Type for Arithmetic Expressions in Stellar

### PR DESCRIPTION
**Pull request by jjmeyer0**
updated stellar arithmetic expressions to be evaluated in a manner that keeps type. For example, if two integers are added then an integer will be returned.

### CODE REVIEW

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/evaluators/ArithmeticEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    final Number l = p.getKey().getValue();
+    final Number r = p.getValue().getValue();
+
+    return function.apply(l == null ? 0 : l, r == null ? 0 : r);
+  }
+
+  public interface ArithmeticEvaluatorFunctions {
+    static BiFunction<Number, Number, Token<? extends Number>> addition() {
+      return (Number l, Number r) -> {
+        if (l instanceof Double || r instanceof Double) {
```
Comment (cestella) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/evaluators/ArithmeticEvaluator.java]: What about if l or r is Float?

**Comment by cestella** [comment]
Oops also we will want to preserve Longs.  Down converting to an int will lose precision and may end up being inaccurate.  I'd suggest we operate like so:
* If either are a double then operate as doubles
* else if either are float then operate as floats
* else if either are longs then operate as longs
* else operate as ints

**Comment by justinleet** [comment]
I agree on operating Casey's rules, given that they match Java's rules for handling this.

http://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.6.2

**Comment by jjmeyer0** [comment]
The reason I neglected longs and floats is because I didn't see the Stellar dsl having the ability to represent them. For example, the test below fails with a parse exception:

`
  @Test
  public void testLongAddition() throws Exception {
    String query = "2147483650 + 2";
    Assert.assertEquals(2147483652L, run(query, new HashMap<>()));
  }
`

That being said it makes sense to still handle longs and floats, if Stellar will at some point. If not that, at the very least if this evaluator sees a long or float it should error out.

Should I add support for longs/floats and then create a Jira for Stellar to add them? Or am I mistaken and Stellar already has support?


Thanks for the feedback.

**Comment by cestella** [comment]
You can return them from as results from stellar functions. Longs are
output from time stamp functions for instance.

File: metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/evaluators/ArithmeticEvaluator.java (hunk 1 above)
Comment (ottobackwards) [inline metron-platform/metron-common/src/main/java/org/apache/metron/common/stellar/evaluators/ArithmeticEvaluator.java]: My only question here, and I may be out of my depth, is does this functionality match the behavior of java itself?  Or other languages?  Would the result of operations on mixed types be consistent (rule of least surprise)?  For example ( and again I'm just imagining here ) Float + Long = Float.    Long + Float = Long.

Hopefully this makes sense

**Comment by ottobackwards** [comment]
[moving this to unbury it]
My only question here, and I may be out of my depth, is does this functionality match the behavior of java itself? Or other languages? Would the result of operations on mixed types be consistent (rule of least surprise)? For example:

 Float + Long = Float
Long + Float = Long

Hopefully this makes sense

**Comment by cestella** [comment]
@ottobackwards Yeah, that was my concern too.  The suggestion I made was based around the Java spec (http://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.6.2 as @justinleet linked earlier).  This is intended to make it follow the rule of least surprise (I think ;).  In your example, Long + Float  should be Float by my suggestion.

**Comment by ottobackwards** [comment]
@cestella I see it now, but not when I made the comment.  Sorry for adding confusion.

+1 by inspection

File: metron-platform/metron-common/src/test/java/org/apache/metron/common/stellar/StellarTest.java [hunk 2]
```diff
@@ -56,6 +56,37 @@
...
+  public void addingDoubleAndIntegerWhereSubjectIsDoubleShouldYieldADouble() throws Exception {
+    String query = "2.1 + 1";
+    Assert.assertEquals(3.1, run(query, new HashMap<>()));
+  }
+
+  @Test
+  public void addingDoubleAndIntegerWhereSubjectIsIntegerShouldYieldADouble() throws Exception {
+    String query = "1 + 2.1";
+    Assert.assertEquals(3.1, run(query, new HashMap<>()));
+  }
```
Comment (cestella) [inline metron-platform/metron-common/src/test/java/org/apache/metron/common/stellar/StellarTest.java]: Could we ensure that we have the following test cases and they return the proper types:
* float + long returns a float
* long + float returns a float
* double + int returns a double
* int + double returns a double
* float + double returns  a double
* double + float returns a double
* long + double returns a double
* double + long returns a double
* long + int returns a long
* int + long returns a long
* int + int returns int
* float + float returns float
* double + double returns double
* long + long returns long

Comment (cestella) [inline metron-platform/metron-common/src/test/java/org/apache/metron/common/stellar/StellarTest.java]: Obviously, the same set for multiplication and division.  I'd parameterize that case ;)

Comment (ottobackwards) [inline metron-platform/metron-common/src/test/java/org/apache/metron/common/stellar/StellarTest.java]: Casey - just like with the stringfunctionstest - I'm not sure that all of this should be in stellartest.  How do you feel about breaking it out?

Comment (cestella) [inline metron-platform/metron-common/src/test/java/org/apache/metron/common/stellar/StellarTest.java]: @ottobackwards Definitely; agreed.  We should have a `StellarArithmeticTest` where we have this.

**Comment by jjmeyer0** [comment]
I believe I have updated all tests and addressed each of the concerns. Thanks again for all the help!

**Comment by cestella** [comment]
Cool!  Dig the tests.  +1 by inspection

**Pull request closed without merging**